fault/verilator_target.py

```diff
@@ -1,5 +1,154 @@
+            if t is None or t._value is None:
+                val = "0"
+                X = "true"
+            else:
+                val = t.as_binary_string()
+                X = "false"
+            return f"{{{val}, {X}}}"
+
+        for t in test:
+            if isinstance(t, Array):
+                testvector.extend(t.flattened())
+            else:
+                testvector.append(t)
+        testvector = ', '.join([to_string(t) for t in testvector])
+        # testvector += ', {}'.format(int(func(*test[:nargs])))
+        source += '''\
+        {{ {} }},
+'''.format(testvector)
+    source += '''\
+    };
+'''
+
+    source += '''
+    for(int i = 0; i < {}; i++) {{
+        value_t* test = tests[i];
+'''.format(len(tests))
+
+    i = 0
+    for name, port in circuit.interface.ports.items():
+        if port.isoutput():
+            if isinstance(port, m.ArrayType):
+                raise NotImplementedError()
+            else:
+                source += f'''\
+        top->{name} = test[{i}].value;
+'''
+        i += 1
+
+    source += '''\
+        top->eval();
+'''
+
+    i = 0
+    for name, port in circuit.interface.ports.items():
+        if port.isinput():
+            source += f'''\
+        check(\"{name}\", top->{name}, test[{i}], i);
+'''
+        i += 1
+    source += '''\
+    }
+'''
+
+    source += '''
+    delete top;
+    std::cout << "Success" << std::endl;
+    exit(0);
+}'''
+
+    return source
+
+
+def compile_verilator_harness(basename, circuit, tests, input_ranges=None):
+    if config.get_compile_dir() == 'callee_file_dir':
+        (_, filename, _, _, _, _) = \
+            inspect.getouterframes(inspect.currentframe())[1]
+        file_path = os.path.dirname(filename)
+        filename = os.path.join(file_path, basename)
+    else:
+        filename = basename
+
+    if callable(tests):
+        tests = testvectors(circuit, tests, input_ranges)
+    verilatorcpp = harness(circuit, tests)
+
+    with open(filename, "w") as f:
+        f.write(verilatorcpp)
+
+
+def run_verilator_test(verilog_file_name, driver_name, top_module,
+                       verilator_flags="", build_dir="build"):
+    if isinstance(verilator_flags, list):
+        if not all(isinstance(flag, str) for flag in verilator_flags):
+            raise ValueError("verilator_flags should be a str or list of strs")
+        verilator_flags = " ".join(verilator_flags)
+    assert not subprocess.call(
+        f'verilator -Wall -Wno-INCABSPATH -Wno-DECLFILENAME {verilator_flags}'
+        f' --cc {verilog_file_name}.v --exe {driver_name}.cpp'
+        f' --top-module {top_module}',
+        cwd=build_dir, shell=True
+    )
+    assert not subprocess.call(
+        f'make -C obj_dir -j -f V{top_module}.mk V{top_module}', cwd=build_dir,
+        shell=True)
+    assert not subprocess.call('./obj_dir/V{}'.format(top_module),
+                               cwd=build_dir, shell=True)
 
 
 class VerilatorTarget(Target):
```

Comment:
Also eventually: we should add a flag as to whether or not the files should be kept around or not. If so, we can put temp files in a dir the user passes in, otherwise can use a temporary dir. The interface could be that if `directory` is None, we put in a tempdir, otherwise in `directory`